File: apps/lantern-api/src/features/identity/presentation/http/handle_auth.py
import json
import re
from typing import Optional
from urllib.parse import parse_qs, urlparse

from src.fireweave.fw_providers import register_fw_target


def _json(data: dict, status: int = 200) -> tuple[int, bytes]:
    return status, (json.dumps(data) + "\n").encode()


def _bearer(headers) -> Optional[str]:
    auth = headers.get("Authorization") or headers.get("authorization") or ""
    m = re.match(r"^Bearer\s+(.+)$", auth, re.I)
    return m.group(1).strip() if m else None


def handle_auth_config(store) -> tuple[int, bytes]:
    return _json({"enabled": store.enabled})


def handle_auth_users(store, path: str) -> tuple[int, bytes]:
    if not store.enabled:
        return _json({"error": "identity disabled"}, 404)
    qs = parse_qs(urlparse(path).query)
    limit_raw = (qs.get("limit") or ["50"])[0]
    try:
        limit = int(limit_raw)
    except ValueError:
        limit = 50
    return _json({"users": store.list_users(limit)})


def handle_auth_session(store, method: str, headers, body: bytes) -> tuple[int, bytes]:
    if not store.enabled:
        return _json({"error": "identity disabled"}, 404)
    if method == "GET":
        sess = store.session(_bearer(headers))
        if not sess:
            return _json({"error": "no session"}, 401)
        return _json(sess)
    if method == "POST":
        try:
            payload = json.loads(body.decode("utf-8") or "{}")
        except json.JSONDecodeError:
            return _json({"error": "invalid json"}, 400)
        user_id = str(payload.get("userId", "")).strip()
        if not user_id:
            return _json({"error": "userId required"}, 400)
        logged = store.login(user_id)
        if not logged:
            return _json({"error": "unknown user"}, 404)
        register_fw_target(
            logged["user"]["id"],
            properties=logged["evaluationContext"]["properties"],
        )
        return _json(
            {
                "sessionToken": logged["token"],
                "user": logged["user"],
                "evaluationContext": logged["evaluationContext"],
            }
        )
    if method == "DELETE":
        return _json({"ok": True})
    return _json({"error": "method not allowed"}, 405)

File: apps/atlas-api/src/features/identity/presentation/http/handle-auth.ts
import type { IdentityStore } from "../../application/ports/identity-store";
import { registerFwTarget } from "../../../../fireweave/fw-providers";

function json(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data) + "\n", {
    status,
    headers: { "content-type": "application/json" },
  });
}

function bearer(req: Request): string | undefined {
  const h = req.headers.get("authorization") ?? "";
  const m = /^Bearer\s+(.+)$/i.exec(h);
  return m?.[1]?.trim();
}

export function handleAuthConfig(store: IdentityStore): Response {
  return json({ enabled: store.enabled });
}

export function handleAuthUsers(store: IdentityStore, req: Request): Response {
  if (!store.enabled) return json({ error: "identity disabled" }, 404);
  const url = new URL(req.url);
  const limit = Number.parseInt(url.searchParams.get("limit") ?? "50", 10);
  return json({ users: store.listUsers(limit) });
}

export async function handleAuthSession(
  store: IdentityStore,
  req: Request,
): Promise<Response> {
  if (!store.enabled) return json({ error: "identity disabled" }, 404);
  if (req.method === "GET") {
    const sess = store.session(bearer(req));
    if (!sess) return json({ error: "no session" }, 401);
    return json(sess);
  }
  if (req.method === "POST") {
    let body: { userId?: string };
    try {
      body = (await req.json()) as { userId?: string };
    } catch {
      return json({ error: "invalid json" }, 400);
    }
    const userId = body.userId?.trim();
    if (!userId) return json({ error: "userId required" }, 400);
    const logged = store.login(userId);
    if (!logged) return json({ error: "unknown user" }, 404);
    // Cohort identity — always-on, never flag-gated (INIT-S8).
    void registerFwTarget(userId, {
      properties: {
        email: logged.user.email,
        name: logged.user.name,
        org: logged.user.org,
        plan: logged.user.plan,
        country: logged.user.country,
      },
    });
    return json({
      sessionToken: logged.token,
      user: logged.user,
      evaluationContext: logged.evaluationContext,
    });
  }
  if (req.method === "DELETE") {
    return json({ ok: true });
  }
  return json({ error: "method not allowed" }, 405);
}
